renku/service/views/datasets.py

```diff
@@ -0,0 +1,325 @@
+              in: query
+              description: Project identifier.
+              schema:
+                type: string
+              required: true
+
+            - name: dataset_id
+              in: query
+              description: Dataset identifier.
+              schema:
+                type: string
+              required: true
+
+        responses:
+            result:
+                description: |
+                  Contains `datasets` key with
+                  list of all datasets from a specified project.
+
+            error:
+                description: |
+                  Contains `code` and `message`
+                  describing the error.
+
+    """
+    if request.method == 'GET':
+        dataset_name = request.args.get('dataset_id')
+        project_id = request.args.get('project_id')
+        pwd = validate_project_id(project_id)
+        if not pwd:
+            return jsonify(status=2, message='invalid project_id argument')
+
+        cwd = os.getcwd()
+        os.chdir(str(pwd))
+
+        files = [
+            ds for ds in
+            json.loads(list_files(dataset_name, None, None, None, 'json-ld'))
+        ]
+
+        os.chdir(cwd)
+        return jsonify(dataset=dataset_name, files=files)
+
+
+@requires_identity
+def add_file_to_dataset_view(user):
+    """
+    Add uploaded file to cloned repository.
+
+    ---
+    get:
+        tags:
+        - "datasets"
+
+        summary: Add a file from cache to cloned repository.
+        description: Add a file from cache to cloned repository
+
+        parameters:
+            - name: dataset_name
+              in: query
+              description: Name of the dataset.
+              schema:
+                type: string
+              required: true
+
+            - name: file_id
+              in: query
+              description: File identifier to add to project.
+              schema:
+                type: string
+              required: true
+
+            - name: project_id
+              in: query
+              description: Project identifier.
+              schema:
+                type: string
+              required: true
+
+            - name: remote_name
+              in: query
+              description: Name of the remote to sync the project with.
+              schema:
+                type: string
+
+        responses:
+            result:
+                description: |
+                  Contains `dataset` and `file_id` keys used in operation.
+
+            error:
+                description: |
+                  Contains `code` and `message` describing the error.
+
+    """
+    if request.method == 'GET':
+        dataset_name = request.args.get('dataset_name')
+        project_id = request.args.get('project_id')
+        file_id = request.args.get('file_id')
+        remote_name = request.args.get('remote_name', 'origin')
+
+        local_path = find_file(user, file_id)
+        if not local_path or not local_path.exists():
```

Comment:
in which case would `local_path.exists()` return `False`?

Reply:
This is just a defensive pattern. The potential case would be: A local path is cached object, a function can return you a path of a file which in that time was in a process of cleanup by a TTL cleanup strategy. During `find_file` processing invocation and found value, cleanup strategy executor can easily delete a file that he found. By the time function returns it can happen that path no longer exists, that's why I added this additional check.